### PR TITLE
Add GraphQL setup

### PR DESCRIPTION
**Pull request by Tom-Bonnike**

### CODE REVIEW

**Comment by PauuloG** [comment]
I don't think we should merge this. See comment on #20.

We should have a proper Relay setup including the standard Relay root query fields (see https://jobs.zalando.com/tech/blog/building-a-relay-compatible-graphql-server/?gh_src=4n3gxh1 for details).

Implementing all relay specs (node resolver, viewer type, and using connections) will enable us to leverage GraphQL in a more efficient way.
I don't see the point of merging an Apollo setup for now. We should discuss this

**Comment by PauuloG** [comment]
Following up on this

After thoroughly reviewing this setup, I have several remarks. I will try to give links and proper documentation for each of the points I make.

### Graphql files

I do not think we should be writing `.graphql` files ourselves. Those files are the transposition of our application's typing system into the GraphQL shorthand typing system. They should be generated automatically for two reasons :
+ Automating this process can help detecting schema errors that we could introduce if we modified the schema ourselves. The schema we expose through our API is public and will be used by all of the consumers (our own and potential others). Breaking the schema is easy, hard to debug without tools, and will break the prod instantly.
+ Generating the schema will enable us to use structures like interfaces that will be automatically generated

In my opinion, we should define types as code, and have them generated automatically to create a single `schema.graphql` file.

See examples of schema defined as code :
+ [graphql-relay-js definition of GraphQL's test schema](https://github.com/graphql/graphql-relay-js/blob/master/src/__tests__/starWarsSchema.js)
+ [Relay fullstack repo graphql schema](https://github.com/lvarayut/relay-fullstack/blob/master/server/data/schema.js)

Those two particular examples don't use stitching, but we could make it work, defining a schema per model and stitching them when generating the `.graphql` files.
I'm leaning toward stitching myself, and I know @tombonnike wants it (let me know what you think).

### Relay specs

Both the examples above implement Relay specs using `graphql-relay-js`, which is a library exposing Relay specific patterns as types and interfaces.

Those patterns can be summarized as such :
#### The Connection interface
The connection system in Relay is designed to help manipulate collections through a GraphQL API.

It comes with a built-in cursor based pagination which makes it easy for the consumer to know at all times how much of the collection he has received, how much is left, and which part of the collection he is currently manipulating.

The connection can be seen as a one to many relationship between different types.
A connection looks like this :
```
# A connection to a list of items.
type UserProjectsConnection {
  # Information to aid in pagination. (type provided by Relay)
  pageInfo: PageInfo!

  # Actual data (an edge type)
  edges: [ProjectsEdge]
}

# An edge in a connection.
type ProjectsEdge {
  # The item at the end of the edge.
  node: Project

  # A cursor for use in pagination. (cursor is the place of our object in the connection)
  cursor: String!
}
```
#### The Node interface
The node interface defines the creation of a unique ID for each object manipulated by the API. This ID is standardized. It takes the application layer arbitrary ID (generally a database ID for example) and the object Type and generates an ID that should be unique across the whole application. Final ID is a base64 encode of TypeName:ArbitraryId

This enables any node to be resolved with a single resolver : the node resolver. The node resolver takes a Relay ID as input and resolves the type of the object first, then the object itself (using the arbitrary ID).
This resolver is very versatile and very useful (when fetching info for a particular element in a list for example).

#### The Root Query
In Relay, the typical root Query should only have two fields, and in GraphQL shorthand type system looks like this :
```
type QueryRoot {
  # Fetches an object given its ID
  node(
    # The ID of an object
    id: ID!
  ): Node

  # The viewer ; May not be connected
  viewer: Viewer
}
```
The Viewer type is conceptually the user that requests the page, which makes permissions management a bit easier.

There are other concepts such as MutationId, but we might not use this for now, and we can still implement this later. Subscriptions are also brought by Relay and enable realtime con

#### Naming conventions
In Relay, especially on the front-end part, there are conventions for typing and naming.
+ Expected user input and API response should be typed, and the names of those types should respectively be MutationNameInput and MutationNamePayload
+ Connections should be named

### Overall

I think we should follow those specs. They might be a bit harder to implement at first, but all the handy stuff will greatly improve the reliability of our app and the strict conventions will make it less hard to contribute to.

I'm in favor of declaring our schema as code in order to leverage the `relay-graphql-js` package which will let us use Relay patterns in our code (cursor system is deeply linked to db queries for example) across the whole app.

**Comment by rmnbrd** [comment]
I mainly agree with @Tom-Bonnike on this one. Relay is very opinionated, and that's a good think. But to be honest, I think Relay lacks of examples and documentation on some topics. 
In my opinion, using Relay is great in a React-oriented environment, but I think it will cause us some pain to work with when it comes to the CLI for exemple.
Moreover, I think using Relay will be a strong entry barrier for some newcomer who would like to contribute to our project.

**TL;DR:**
Should we use Apollo?

**Comment by PauuloG** [comment]
If both of you prefer to use Apollo it's fine. 

A few things however: 
>  If I’m not mistaken, Relay was built to be used with React only…
> One could also argue that because we use Relay, our consumers will HAVE to use a relay-compliant client
Relay are just specs on top of GraphQL, so any GraphQL client should be able to use a Relay Server just as well a any other. Front-end Relay helps leveraging all Relay server Specs but (in the GraphQL spirit really), in the end it's all just specs that you can chose to use or not.

> Server-side rendering support is also not as great 
I might be wrong but I don't think we will need that 

> I think Relay lacks of examples and documentation on some topics.
That's true especially on the front-end side. Examples exist but they are sometimes hard to find.

> Should we use Apollo?
I mean yeah, if you want to.

**Comment by Tom-Bonnike** [comment]
> Relay are just specs on top of GraphQL, so any GraphQL client should be able to use a Relay Server just as well a any other. Front-end Relay helps leveraging all Relay server Specs but (in the GraphQL spirit really), in the end it's all just specs that you can chose to use or not.

I get that. But we’d still be forcing ourselves into respecting opinionated specs that we won’t benefit from in our CLI and that might make our lives harder. For example, we can build our pagination system the way we want if we ever need to adapt it for our CLI.
We can totally cherry-pick some of Relay’s rules, though.

File: src/server/index.js [hunk 1]
```diff
@@ -14,6 +17,14 @@ export default (): express$Application => {
 
   // Routes
   app.get('/', indexRoute)
+  app.use(GRAPHQL_ENDPOINT, graphqlExpress({ schema }))
+
+  if (process.env.NODE_ENV !== 'production') {
+    app.get(
+      GRAPHIQL_ENDPOINT,
+      graphiqlExpress({ endpointURL: GRAPHQL_ENDPOINT })
```
Comment (Tom-Bonnike) [inline src/server/index.js]: This should be `GRAPHIQL_ENDPOINT`
